### PR TITLE
T2D-245 Support for reference gene sequences

### PR DESCRIPTION
**Pull request by iyangar**
1. Existing Assembly is renamed to ReferenceSequence and all dependencies, variables are updated.
2. New type is introduced to specify Ref Seq -> Assembly, Gene or Transcriptome
3. Test cases are added
4. Swagger UI is enhanced

Includes  T2D-246: Supporting transcriptome

### CODE REVIEW

File: src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/ReferenceSequenceRestController.java [hunk 1]
```diff
@@ -33,44 +33,47 @@
...
-    @ApiOperation(value="Get a filtered list of assemblies based on filtering criteria")
+    @ApiOperation(value="Get a filtered list of referenceSequences based on filtering criteria")
     @ApiImplicitParams({
             @ApiImplicitParam(name = "name", value = "name", dataType = "string", paramType = "query", example = "GRCh38"),
             @ApiImplicitParam(name = "patch", value = "patch number", dataType = "string", paramType = "query", example = "p2"),
-            @ApiImplicitParam(name = "accessions", value = "accession", dataType = "string", paramType = "query", example = "GCA_000001405.3")
+            @ApiImplicitParam(name = "accessions", value = "accession", dataType = "string", paramType = "query", example = "GCA_000001405.3"),
+            @ApiImplicitParam(name = "type", value = "ReferenceSequence's type", dataType = "string",
+                    paramType = "query", example = "ASSEMBLY",
+                    allowableValues = "ASSEMBLY,GENE,TRANSCRIPTOME")
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/ReferenceSequenceRestController.java]: These parameters can fit in 2 lines.

File: src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java [hunk 2]
```diff
@@ -118,20 +118,24 @@ public void cleanDatabases() throws Exception {
     }
 
     @Test
-    public void postAssembly() throws Exception {
-        String location = postTestAssembly("GRCh37", "p2",
+    public void postReferenceSequence() throws Exception {
```
Comment (cyenyxe) [inline src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java]: Please add tests for other reference sequence types.

File: src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Analysis.java [hunk 3]
```diff
@@ -90,10 +90,10 @@
     @ManyToOne(optional = false)
     private Study study;
 
-    @ApiModelProperty(position = 5, dataType = "java.lang.String", notes = "Url to an Assembly")
+    @ApiModelProperty(position = 5, dataType = "java.lang.String", notes = "Url to an ReferenceSequence")
```
[the remark this answers is not in the document]
Comment (iyangar) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Analysis.java]: Currently we have "Url to a **S**tudy" and "Url to a **T**axonomy" with entity names starting with Capital letters. Let me know if `Reference Sequence` or `reference sequence` is better

[the remark this answers is not in the document]
Comment (cyenyxe) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Analysis.java]: As a descriptive string, I would personally prefer the lowercase. Definitely 'Url' should be 'URL'.

File: src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/ReferenceSequenceRestController.java [hunk 4]
```diff
@@ -33,44 +33,47 @@
...
+import uk.ac.ebi.ampt2d.metadata.rest.resources.ReferenceSequenceResource;
 
 import java.util.List;
 
 @RestController
-@Api(tags = "Assembly Entity")
-@RequestMapping(path = "assemblies")
-public class AssemblyRestController implements ResourceProcessor<RepositoryLinksResource> {
+@Api(tags = "ReferenceSequence Entity")
+@RequestMapping(path = "referenceSequences")
```
[the remark this answers is not in the document]
Comment (iyangar) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/ReferenceSequenceRestController.java]: Sure, we would refer without  ```-``` and only in the URLs will have reference-sequences.

[the remark this answers is not in the document]
Comment (iyangar) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/ReferenceSequenceRestController.java]: Changing the URL to '/reference-sequences' works for search but regular get/post still expects /referenceSequecesis. It appears Swagger is overriding it. Hence comment not addressed as URL would be different for different operations.

[the remark this answers is not in the document]
Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/ReferenceSequenceRestController.java]: I guess urls are case sensitive but you can change the path to required ones by overriding path in  @RepositoryRestResource(path="requiredUrl")

File: src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/ReferenceSequenceRestController.java [hunk 5]
```diff
@@ -33,44 +34,46 @@
...
-        Resources<AssemblyResource> resources = (Resources<AssemblyResource>) resourceAssembler.toResources(Assembly.class, assemblies);
+        Resources<ReferenceSequenceResource> resources = (Resources<ReferenceSequenceResource>) resourceAssembler.toResources(ReferenceSequence.class, referenceSequences);
 
         return ResponseEntity.ok(resources);
     }
 
     @Override
     public RepositoryLinksResource process(RepositoryLinksResource resource) {
-        resource.add(ControllerLinkBuilder.linkTo(AssemblyRestController.class).slash("/search").withRel("assemblies"));
+        resource.add(ControllerLinkBuilder.linkTo(ReferenceSequenceRestController.class).slash("/search").withRel("referenceSequences"));
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/ReferenceSequenceRestController.java]: Can you also change this to `reference-seequences` please?

Comment (iyangar) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/ReferenceSequenceRestController.java]: Done.
Looked another place during testing and working on it

**Pull request merged**